core/remote_state_verifier.go

```diff
@@ -0,0 +1,438 @@
+	verifyTaskFailedMeter  = metrics.NewRegisteredMeter("verifymanager/task/result/failed", nil)
+
+	verifyTaskExecutionTimer = metrics.NewRegisteredTimer("verifymanager/task/execution", nil)
+)
+
+type remoteVerifyManager struct {
+	bc            *BlockChain
+	taskLock      sync.RWMutex
+	tasks         map[common.Hash]*verifyTask
+	peers         verifyPeers
+	verifiedCache *lru.Cache
+	allowInsecure bool
+
+	// Subscription
+	chainBlockCh chan ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	// Channels
+	verifyCh  chan common.Hash
+	messageCh chan verifyMessage
+}
+
+func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowInsecure bool) (*remoteVerifyManager, error) {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	block := blockchain.CurrentBlock()
+	if block == nil {
+		return nil, ErrCurrentBlockNotFound
+	}
+
+	// rewind to last non verified block
+	number := new(big.Int).Sub(block.Number(), big.NewInt(int64(maxForkHeight)))
+	if number.Cmp(common.Big0) < 0 {
+		blockchain.SetHead(0)
+	} else {
+		numberU64 := number.Uint64()
+		blockchain.SetHead(numberU64)
+		block := blockchain.GetBlockByNumber(numberU64)
+		for i := 0; i < maxForkHeight && block.NumberU64() > 0; i++ {
+			// When inserting a block,
+			// the block before 11 blocks will be verified,
+			// so the parent block of 11-22 will directly write the verification information.
+			verifiedCache.Add(block.Hash(), true)
+			block = blockchain.GetBlockByHash(block.ParentHash())
+			if block == nil {
+				return nil, fmt.Errorf("block is nil, number: %d", number)
+			}
+		}
+	}
+
+	vm := &remoteVerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*verifyTask),
+		peers:         peers,
+		verifiedCache: verifiedCache,
+		allowInsecure: allowInsecure,
+
+		chainBlockCh: make(chan ChainHeadEvent, chainHeadChanSize),
+		verifyCh:     make(chan common.Hash, maxForkHeight),
+		messageCh:    make(chan verifyMessage),
+	}
+	vm.chainHeadSub = blockchain.SubscribeChainBlockEvent(vm.chainBlockCh)
+	return vm, nil
+}
+
+func (vm *remoteVerifyManager) mainLoop() {
+	defer vm.chainHeadSub.Unsubscribe()
+
+	pruneTicker := time.NewTicker(pruneInterval)
+	defer pruneTicker.Stop()
+	for {
+		select {
+		case h := <-vm.chainBlockCh:
+			vm.NewBlockVerifyTask(h.Block.Header())
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			vm.taskLock.Lock()
+			if task, ok := vm.tasks[hash]; ok {
+				delete(vm.tasks, hash)
+				verifyTaskCounter.Dec(1)
+				verifyTaskSucceedMeter.Mark(1)
+				verifyTaskExecutionTimer.Update(time.Since(task.startAt))
+				task.Close()
+			}
+			vm.taskLock.Unlock()
+		case <-pruneTicker.C:
+			vm.taskLock.Lock()
+			for hash, task := range vm.tasks {
+				if vm.bc.insertStopped() || (vm.bc.CurrentHeader().Number.Cmp(task.blockHeader.Number) == 1 &&
+					vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > pruneHeightDiff) {
+					delete(vm.tasks, hash)
```

Comment:
```
delete(vm.tasks, hash)
verifyTaskCounter.Dec(1)
verifyTaskFailedMeter.Mark(1)
task.Close()
```
Can be a function to reduce duplicated code.

Reply:
> ```
> delete(vm.tasks, hash)
> verifyTaskCounter.Dec(1)
> verifyTaskFailedMeter.Mark(1)
> task.Close()
> ```
> 
> Can be a function to reduce duplicated code.

fixed

Reply:
1. `CloseTaskAndMarkMeters` ==> `CloseTask` is enough, we don't have to reflect everything on the function name.
2.  We should put `verifyTaskExecutionTimer.Update(time.Since(task.startAt))` inside `CloseTask`.
3.  Two parameter `hash common.Hash, task *verifyTask` are duplicated for `CloseTask`, `task` is good enough, we can get `hash` from `task`, dont we?

Reply:
> 1. `CloseTaskAndMarkMeters` ==> `CloseTask` is enough, we don't have to reflect everything on the function name.
> 2. We should put `verifyTaskExecutionTimer.Update(time.Since(task.startAt))` inside `CloseTask`.
> 3. Two parameter `hash common.Hash, task *verifyTask` are duplicated for `CloseTask`, `task` is good enough, we can get `hash` from `task`, dont we?

fixed